cmd/utils/nodecmd/snapshot.go

```diff
@@ -0,0 +1,272 @@
+
+	snaptree, err := snapshot.New(db, statedb.NewDatabase(db), 256, headBlock.Root(), false, false, false)
+	if err != nil {
+		logger.Error("Failed to open snapshot tree", "err", err)
+		return err
+	}
+	if ctx.NArg() > 1 {
+		logger.Error("Too many arguments given")
+		return errors.New("too many arguments")
+	}
+	root := headBlock.Root()
+	if ctx.NArg() == 1 {
+		root, err = parseRoot(ctx.Args().First())
+		if err != nil {
+			logger.Error("Failed to resolve state root", "err", err)
+			return err
+		}
+	}
+	if err := snaptree.Verify(root); err != nil {
+		logger.Error("Failed to verify state", "root", root, "err", err)
+		return err
+	}
+	logger.Info("Verified the state", "root", root)
+	return nil
+}
+
+func traceTrie(ctx *cli.Context) error {
+	stack := MakeFullNode(ctx)
+	dbm := stack.OpenDatabase(getConfig(ctx))
+	head := dbm.ReadHeadBlockHash()
+	if head == (common.Hash{}) {
+		// Corrupt or empty database, init from scratch
+		return errors.New("empty database")
+	}
+	// Make sure the entire head block is available
+	tmpHeadBlock := dbm.ReadBlockByHash(head)
+	if tmpHeadBlock == nil {
+		return fmt.Errorf("tmp head block missing: %v", head.String())
+	}
+
+	blockNumber := (tmpHeadBlock.NumberU64() / 128) * 128
+	headBlock := dbm.ReadBlockByNumber(blockNumber)
+	if headBlock == nil {
+		return fmt.Errorf("head block missing: %v", head.String())
+	}
+
+	root := headBlock.Root()
+	if root == (common.Hash{}) {
+		// Corrupt or empty database, init from scratch
+		return errors.New("empty root")
+	}
+
+	logger.Info("Trace Start", "BlockNum", blockNumber)
+
+	sdb, err := state.New(root, state.NewDatabase(dbm), nil)
+	if err != nil {
+		return fmt.Errorf("Failed to open newDB trie : %v", err)
+	}
+	tridb := sdb.Database().TrieDB()
+
+	children, err := tridb.NodeChildren(root)
+	if err != nil {
+		return fmt.Errorf("Fail get childrens of root : %v", err)
+	}
+
+	var wait sync.WaitGroup
+	wait.Add(len(children))
+	for _, child := range children {
+		go func(eth common.Hash) {
+			defer wait.Done()
+			state.Tracer(sdb.Database(), eth)
+		}(child)
+	}
+
+	wait.Wait()
+	logger.Info("Trie Tracer finished")
+	return nil
+}
+
+func iterateTrie(ctx *cli.Context) error {
+	stack := MakeFullNode(ctx)
+	dbm := stack.OpenDatabase(getConfig(ctx))
+	sdb, err := state.New(common.Hash{}, state.NewDatabase(dbm), nil)
+	if err != nil {
+		return fmt.Errorf("Failed to open newDB trie : %v", err)
+	}
+
+	logger.Info("TrieDB Iterator Start", "node count : all node count, nil node count : key or value is nil node count")
+	Cnt, nilCnt := uint64(0), uint64(0)
+	go func() {
+		for {
+			time.Sleep(time.Second * 5)
+			logger.Info("TrieDB Iterator", "node count", Cnt, "nil node count", nilCnt)
+		}
+	}()
+
+	it := sdb.Database().TrieDB().DiskDB().GetStateTrieDB().NewIterator(nil, nil)
+	for it.Next() {
+		Cnt++
+		if it.Key() == nil || it.Value() == nil || bytes.Equal(it.Key(), []byte("")) || bytes.Equal(it.Value(), []byte("")) {
+			nilCnt++
+		}
+	}
+	it.Release()
```

Comment:
In other code references, Iterator uses defer Release(). How about moving the 'defer it.Release()' statement to appear right after the 'it' variable is assigned.